src/vehicle.cpp

```diff
@@ -1139,13 +1156,31 @@ void ViewportAddVehicles(DrawPixelInfo *dpi)
 			const Vehicle *v = _vehicle_viewport_hash[x + y]; // already masked & 0xFFF
 
 			while (v != nullptr) {
+
 				if (!(v->vehstatus & VS_HIDDEN) &&
 						l <= v->coord.right &&
 						t <= v->coord.bottom &&
 						r >= v->coord.left &&
 						b >= v->coord.top) {
 					DoDrawVehicle(v);
 				}
+				else {
+					/*
+					 * Indicate that this vehicle was considered for rendering in a viewport,
+					 * and we therefore need to update sprites more frequently in case a callback
+					 * will change the bounding box to one which will cause the sprite to be
+					 * displayed.
+					 *
+					 * This reduces the chances of flicker when sprites enter the screen, if they
+					 * are part of a newgrf vehicle set which changes bounding boxes within a
+					 * single vehicle direction.
+					 *
+					 * TODO: is there a cleaner solution than casting to a mutable type?
```

Comment:
If you change `sprite_seq` and `rstate` to be declared with the keyword `mutable`, you don't need to cast. It's a tiny cheat, though.

Maybe `sprite_seq` should be inside `rstate` then, which then becomes a kind `sprite_cache`.

Reply:
Agreed, this would be the better solution